### PR TITLE
INTYGFV-16077: Added failure handler for saml2 login.

### PR DESCRIPTION
**Pull request by ChrilleSigerhed**

### CODE REVIEW

File: app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java [hunk 1]
```diff
@@ -50,6 +53,7 @@ public class WebSecurityConfig {
   public static final String TESTABILITY_API = "/api/testability/**";
   public static final String HEALTH_CHECK_ENDPOINT = "/actuator/health";
   public static final String APP_BUNDLE_NAME = "app";
+  private static final String ERROR_LOGIN = "/error/login/";
```
Comment (mww-sopra-steria) [inline app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java:56]: Maybe call it ERROR_LOGIN_URL

File: app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java [hunk 2]
```diff
@@ -140,6 +145,11 @@ public SecurityFilterChain filterChain(HttpSecurity http,
     return http.build();
   }
 
+  public AuthenticationFailureHandler getForwardAuthenticationFailureHandler() {
+    final var id = String.valueOf(UUID.randomUUID());
```
Comment (mww-sopra-steria) [inline app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java:149]: Won't this create an id that will be used every time anyone fails to login. The randomise should be done on each failed attempt.

Also we need to do log the error so we can find the errorId in our server logs.

Maybe consider to create a custom failure handler that do these things:
1) generate errorId
2) log error using MonitoringLogService
3) redirect the user to /error/login/{errorId}

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Sweet! Nice to introduce a failure handler. We need to consider a few things in the failure handler.

File: app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java (hunk 2 above)
Comment (ChrilleSigerhed) [reply app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java]: You are correct, the previous implementation would only generate one errorId that would be used for all failures.

I updated the solution and instead used a CustomAuthenticationFailureHandler to handle failures, that generates an errorId, log & redirects the user to the correct endpoint.

This log is new, if we decide to use it, will we need to update indexes on elastic?

File: app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java [hunk 3]
```diff
@@ -140,6 +144,10 @@ public SecurityFilterChain filterChain(HttpSecurity http,
     return http.build();
   }
 
+  public AuthenticationFailureHandler getCustomAuthenticationFailureHandler() {
```
Comment (mww-sopra-steria) [inline app/src/main/java/se/inera/intyg/minaintyg/config/WebSecurityConfig.java:147]: Isn't this method redundant. Line 123 could just use the member.

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
1) Log error within CustomAuthenticationFailureHandler
2) Log CLIENT_ERROR
3) Log CITIZEN_LOGIN_FAILURE
4) ERROR_LOGIN_URL as configuration in application.properties

**Review by mww-sopra-steria — APPROVED** [review]
Very nice!

**Pull request merged**